jenkins/JenkinsFile-rebuild-docker-images

```diff
@@ -0,0 +1,318 @@
+// under the License.
+
+// Command to rebuild a docker container
+docker_build = 'docker/build.sh'
+
+// Global variable to store the Docker tag generated
+// by this build.
+TVM_DOCKER_TAG=""
+
+
+def per_exec_ws(folder) {
+  return "workspace/exec_${env.EXECUTOR_NUMBER}/" + folder
+}
+
+
+def upload_to_docker_hub(image_type) {
+  // 'docker/build.sh' makes some assumptions on the images produced.
+  // The expected name is composed by ${BUILD_TAG}.${image_type}:${TVM_DOCKER_TAG}
+
+  sh """
+    docker login -u ${DOCKERHUB_USER} -p ${DOCKERHUB_KEY}
+
+    docker tag \
+      ${BUILD_TAG}.${image_type}:${TVM_DOCKER_TAG} \
+      ${DOCKERHUB_USER}/${image_type}:${TVM_DOCKER_TAG}
+
+    docker push ${DOCKERHUB_USER}/${image_type}:${TVM_DOCKER_TAG}
+  """
+
+}
+
+
+def cleanup_docker_image(image_name) {
+  // Remove a docker image from the local machine.
+  // With the use of -f, makes it not fail in case the image
+  // does not exist.
+
+  sh """
+    docker rmi -f ${image_name}
+  """
+}
+
+
+def init_git() {
+  // Add more info about job node
+  sh """
+     echo "INFO: NODE_NAME=${NODE_NAME} EXECUTOR_NUMBER=${EXECUTOR_NUMBER}"
+     """
+
+  sh "git clone ${TVM_GIT_REPO_URL}"
+
+  dir('tvm') {
+    script{
+      // At the beginning of a build, the $TVM_CURRENT_REF is set
+      // so that all images are generated using strictly the same
+      // TVM git revision.
+      sh "git checkout ${TVM_CURRENT_REF}"
+    }
+  } // dir
+}
+
+
+pipeline {
+  agent { node { label 'CPU' } }
+
+//  options {
+      //timestamps()
+  //}
+
+  environment {
+      DOCKERHUB_USER = "tlcpackstaging"
+      DOCKERHUB_KEY = credentials('dockerhub-tlcpackstaging-key')
+      DISCORD_WEBHOOK = credentials('discord-webhook-url')
+  }
+
+  parameters {
+    string(name:"TVM_GIT_REV",
+           defaultValue: "main",
+           description: 'Git revision to checkout.')
+  }
+
+  stages {
+
+    stage('Prepare') {
+      agent { node { label 'CPU' } }
+      steps {
+        ws(per_exec_ws("tvm/docker-prepare")) {
+          cleanWs();
+          sh "git clone ${TVM_GIT_REPO_URL}"
+          dir('tvm'){
+            script{
+              TVM_CURRENT_REF = sh (
+                script: 'git rev-parse HEAD',
+                returnStdout: true
+              ).trim()
+
+              TIMESTAMP_TAG = sh (
+                script: 'date "+%Y_%m_%d-%H%M%S"',
+                returnStdout: true
+              ).trim()
+
+              TVM_CURRENT_SHORT_REF = sh (
+                script: 'git rev-parse --short HEAD',
+                returnStdout: true
+              ).trim()
+
+              currentBuild.displayName = "${TVM_CURRENT_SHORT_REF}"
+              TVM_DOCKER_TAG = "${TIMESTAMP_TAG}-${TVM_CURRENT_SHORT_REF}"
+              
+            }
+          }
+        }
+      } // steps
+      post {
+        always {
+          cleanWs();
+        }
+      } // post
+    } // stage: Prepare
+
+    stage('Build') {
+      parallel {
+
+        stage('ci-cpu') {
```

Comment:
just one question, is it possible to define these using e.g. a for loop over a dict mapping image name to node label?

Reply:
It would be cool, but I couldn't find a way to do it. We could do that in pure Groovy with a shared library I think, but that open a whole new way of dealing with the jobs, so I recommend we stick with a plain Jenkinsfile like this one.